rskj-core/src/main/java/org/ethereum/listener/GasWeightedCalc.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2017 RSK Labs Ltd.
+ * (derived from ethereumJ library, Copyright (c) 2016 <ether.camp>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.ethereum.listener;
+
+import co.rsk.core.Coin;
+import co.rsk.remasc.RemascTransaction;
+import org.ethereum.core.Block;
+import org.ethereum.core.Transaction;
+import org.ethereum.core.TransactionReceipt;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.math.BigDecimal;
+import java.util.*;
+
+public class GasWeightedCalc implements GasCalculator {
+    private static final Logger logger = LoggerFactory.getLogger("gaspricetracker");
+    private static final int WINDOW_SIZE = 512;
+    private final Deque<GasEntry> gasWindow = new ArrayDeque<>(WINDOW_SIZE);
+    private final Map<Coin, Long> windowMap = new HashMap<>();
```

Comment:
didn't quite get the idea of using two collections. Isn't it enough to have let's say one `LinkedList<GasEntry>` (a queue with max size of 512 with entries corresponding to recent txs that were mined)? - so when a new tx arrives, a new entry of gasPrice/gasUsed is then get added to this collection and oldest one gets removed. Then you can use that list to calculate average value based on txs' costs (gasPrice * gasUsed). Am I missing something in this alg?

Reply:
I'm choose the map to accumulate the amount of gas per gas price. So as the price is being repeated along the TX the calculation is cheaper using the map. The list is used to know the amount of gas to remove from the map. 

In the first iteration I was calculating the value each time it was requested so it was cheaper to use the map but maybe as it is now it is not worthy. 
WDYT?